pika/adapters/blocking_connection.py

```diff
@@ -699,8 +699,10 @@ def process_data_events(self, time_limit=0):
             until I/O produces actionalable events. Defaults to 0 for backward
             compatibility. This parameter is NEW in pika 0.10.0.
         """
-        common_terminator = lambda: bool(
-            self._channels_pending_dispatch or self._ready_events)
+        with self._acquire_event_dispatch() as dispatch_acquired:
+            # Check if we can actually process pending events
+            common_terminator = lambda: bool(dispatch_acquired and
+                (self._channels_pending_dispatch or self._ready_events))
 
         if time_limit is None:
```

Comment:
The context manager `_aquire_event_dispatch()` will yield an integer that is indicating if the resource was acquired (in case > 0).